### PR TITLE
Do not allow send to non-self addresses by default

### PR DESCRIPTION
**Pull request by bvbfan**

### CODE REVIEW

**Comment by monstrobishi** [comment]
oh I see, so sending to own addresses is allowed by default, so this shouldn't cause any issues with the app internally

File: src/masternodes/mn_rpc.cpp [hunk 1]
```diff
@@ -486,6 +486,26 @@ static std::vector<CTxIn> GetAuthInputsSmart(CWallet* const pwallet, int32_t txV
     return result;
 }
 
+static CAccounts DecodeRecipientsDefaultInternlal(CWallet* const pwallet, UniValue const& values) {
```
Comment (monstrobishi) [inline src/masternodes/mn_rpc.cpp:489]: typo internlal

File: src/masternodes/masternodes.cpp [hunk 2]
```diff
@@ -618,17 +620,34 @@ bool CRewardsHistoryStorage::Flush()
     return CCustomCSView::Flush();
 }
 
-isminetype IsMineCached(CWallet const & wallet, CScript const & script)
+struct CCacheInfo
```
Comment (monstrobishi) [inline src/masternodes/masternodes.cpp:623]: what are these changes for? is it an optimisation? will this speed up indexing too?

File: src/masternodes/mn_rpc.cpp [hunk 3]
```diff
@@ -2208,6 +2228,7 @@ UniValue utxostoaccount(const JSONRPCRequest& request) {
                         {
                             {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "The defi address is the key, the value is amount in amount@token format. "
                                                                                  "If multiple tokens are to be transferred, specify an array [\"amount1@t1\", \"amount2@t2\"]"},
+                            {"enable_external_dfi_token_tx", RPCArg::Type::BOOL, RPCArg::Optional::OMITTED, "By default we don't allow sending tokens outside dfi addresses"},
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2231]: The parameter is added but never used?

File: src/masternodes/masternodes.cpp (hunk 2 above)
Comment (ShengguangXiao) [reply src/masternodes/masternodes.cpp:623]: I guess the change is to dynamically update the cache if the wallet changed or unloaded.

**Comment by ShengguangXiao** [comment]
Can add unit test for `enable_external_dfi_token_tx` is false?

File: src/masternodes/masternodes.cpp (hunk 2 above)
Comment (bvbfan) [reply src/masternodes/masternodes.cpp:623]: Update mine cache when you insert private key, also clears cache when wallet is unloaded.

**Comment by bvbfan** [comment]
> Can add unit test for `enable_external_dfi_token_tx` is false?

When it's not specified it's false, most of cases.

**Comment by bvbfan** [comment]
Fix typo.

**Review by ShengguangXiao — APPROVED** [review]

**Comment by monstrobishi** [comment]
tests passing I will merge

**Comment by uzyn** [comment]
This is a bad merge, please rollback this. 

- Bad interface design: it sholdn't be part of the `to` obj.
- And why are we blocking general token send. The original issue was only for DFI token specifically.

**Pull request merged**